pkg/ibmvpc/subnetsIPBlocks.go

```diff
@@ -0,0 +1,252 @@
+//      splitByFiltersBlocks are such that:
+//   		a. the blocks are disjoint
+//	    	b. the union of the blocks is the subnet cidr
+//      we calculate splitByFiltersBlocks of a subnet by:
+//          (a) collecting all the filters' rules blocks.
+//          (b) adding allCidr block to these blocks.
+//          (c) from these blocks, create a list of disjoint blocks
+//          (d) for each of these blocks, intersect the block with the subnet cidr
+//          (e) collect all the non-empty intersections we get in (d)
+//   2. allocate a free address for the fake private IPs - to this end, we hold for each subnet freeAddressesBlocks,
+//      freeAddressesBlocks are splitByFiltersBlocks minus all the addresses that were already allocated.
+//      to get freeAddressesBlocks, we first copy splitByFiltersBlocks, then we remove the subnet already allocated addresses.
+//      (we get these addresses from the subnet reserved IP).
+//       when a free address is needed, we take the first address of the block and remove it from the free blocks list.
+
+// /////////////////////////////////////////////////////////////////////////////////////////
+// subnetIPBlocks hold the blocks of a subnet
+type subnetIPBlocks struct {
+	subnetOriginalBlock  *ipblock.IPBlock // the block of the original cidr of the subnet
+	splitByFiltersBlocks []*ipblock.IPBlock
+	freeAddressesBlocks  []*ipblock.IPBlock // each block in splitByFiltersBlocks has a corresponding block at freeAddressesBlocks
+	fullyReservedBlocks  []bool             // true if all the addresses in the block are reserved IP
+}
+type subnetsIPBlocks map[string]*subnetIPBlocks
+
+func getSubnetsIPBlocks(rc *datamodel.ResourcesContainerModel) (subnetsBlocks subnetsIPBlocks, err error) {
+	subnetsBlocks = subnetsIPBlocks{}
+	// get all the original blocks of the subnets:
+	if err := subnetsBlocks.getSubnetsOriginalBlocks(rc); err != nil {
+		return nil, err
+	}
+	// calc the filters blocks:
+	filtersBlocks, err := getFiltersBlocks(rc)
+	if err != nil {
+		return nil, err
+	}
+	// calc the splitByFiltersBlocks:
+	subnetsBlocks.splitSubnetsOriginalBlocks(rc, filtersBlocks)
+	// calc the freeAddressesBlocks:
+	if err := subnetsBlocks.getSubnetsFreeBlocks(rc); err != nil {
+		return nil, err
+	}
+	return subnetsBlocks, nil
+}
+
+func (subnetsBlocks subnetsIPBlocks) getSubnetsOriginalBlocks(rc *datamodel.ResourcesContainerModel) (err error) {
+	for _, subnetObj := range rc.SubnetList {
+		subnetsBlocks[*subnetObj.CRN] = &subnetIPBlocks{}
+		subnetsBlocks[*subnetObj.CRN].subnetOriginalBlock, err = ipblock.FromCidr(*subnetObj.Ipv4CIDRBlock)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// splitSubnetsOriginalBlocks():
+// the goal of this func is to split the original subnet cidr to disjoint blocks, according to the filters blocks:
+// for each subnet it gets all the blocks that intersect with subnet original block
```

Comment:
Either repeat the accurate description you have above, in the beginning of the file, or refer to it. The description here is not clear enough and may even confuse.

Reply:
fixed

Reply:
You refer to the names of the vars; either explain what each represents or imply that the functionality of each is defined above

Reply:
fixed